File: 프로그래머스/Level1/신규아이디추천/id_recommend.py
"""
첫 번째 풀이에는 함수를 모두 구현하였음에도 중구난방식이라 어디가 어느 기능을 담당하는지 한눈에 보기 힘들었다. 하지만 각각의 기능을 모두 함수로 바꿔 표현하니
훨씬 간단해 보이는 것을 느꼈다.
"""
'''
첫번째 풀이
'''
def recom(new_id):
    #1
    new_id = new_id.lower()
    #2
    for i in new_id:
        if i.islower() == False and i != '-' and i != '_' and i != '.' and i.isnumeric() == False:
            new_id = new_id.replace(i,'')
    #3
    while '..' in new_id:
        new_id = new_id.replace('..','.')
    #4
    if new_id.startswith('.') == True:
        new_id = new_id[1:]
    
    if new_id.endswith('.') == True:
        new_id = new_id[:len(new_id)-1]
    #5   
    if new_id == "":
        new_id += "a"
    #6
    if len(new_id) >= 16:
        new_id = new_id[:15]
        if new_id.endswith('.') == True:
            new_id =new_id[:len(new_id)-1]
    #7
    if len(new_id) <= 2:
        new_id = new_id + new_id[-1]*(3-len(new_id))
        
    return new_id

def solution(new_id):
    answer = new_id
    if len(new_id) < 3 or len(new_id) > 15:
        return recom(new_id)
    
    for i in new_id:
        if i.islower() == False and i != '-' and i != '_' and i != '.' and i.isnumeric() == False:
            return recom(new_id)
    
    if new_id[0] == '.' or new_id[-1] == '.' or '..' in new_id:
        return recom(new_id)
    
    return answer

'''
두번째 풀이
'''


def first(new_id):
    return new_id.lower()


def second(new_id):
    sc = ['-', '_', '.']
    str_temp = ''
    for char in new_id[:]:
        if char.islower() or char.isdigit() or char in sc:
            str_temp += char
    return str_temp


def third(new_id):
    while '..' in new_id:
        new_id = new_id.replace("..", ".")
    return new_id


def fourth(new_id):
    return new_id.strip('.')


def fifth(new_id):
    if not new_id:
        return 'a'
    else:
        return new_id


def sixth(new_id):
    if len(new_id) >= 16:
        return new_id[:15].strip('.')
    else:
        return new_id


def seventh(new_id):
    while len(new_id) <= 2:
        new_id += new_id[-1]
    return new_id


def solution(new_id):
    answer = ''
    new_id = first(new_id)
    new_id = second(new_id)
    new_id = third(new_id)
    new_id = fourth(new_id)
    new_id = fifth(new_id)
    new_id = sixth(new_id)
    new_id = seventh(new_id)
    answer = new_id
    return answer

File: 프로그래머스/Level1/신규아이디추천/id_recommend.js
// 첫 번째 풀이 -> 아직 파이썬답게 풀어 chaining과 js 메소드를 완벽히 사용하지 못하였다.
function first(id){
  return id.toLowerCase();
}

function second(id){
  const reg = /[^0-9a-z._-]/g;
  return id.replace(reg,"");
}

function third(id){
  const reg = /\.+/g;
  id = id.replace(reg,"\.");
  return id
}

function fourth(id){
    const reg = /^\.|\.$/;
    while(reg.test(id)){
        id = id.replace(reg,"");
    }
    return id
}

function fifth(id){
    if (!id){
        return "a";
    }
    return id;
}

function sixth(id){
    if (id.length >= 16){
        id = id.slice(0,15);
    }
    id = fourth(id);
    return id;
}

function seventh(id){
    while(id.length <= 2){
        id += id[id.length-1];
    }
    return id;
}
  
function solution(new_id) {
  var answer = '';
  answer = first(new_id);
  answer = second(answer);
  answer = third(answer);
  answer = fourth(answer);
  answer = fifth(answer);
  answer = sixth(answer);
  answer = seventh(answer);
  console.log(answer);
  return answer;
}

// 다른 사람들의 chaining을 보면서 공부한 메소드를 사용한 콛
function solution(new_id) {
  const answer = new_id
      .toLowerCase()
      .replace(/[^\w-.]/g,"")
      .replace(/\.+/g,".")
      .replace(/^\.|\.$/, "")
      .replace(/^$/,"a")
      .slice(0,15).replace(/\.$/, "");
  return answer.padEnd(3, answer[answer.length-1]);
}